pkg/querier/batch/merge.go

```diff
@@ -52,6 +61,29 @@ func newMergeIterator(cs []GenericChunk) *mergeIterator {
 	return c
 }
 
+func (c *mergeIterator) Reset(size int) *mergeIterator {
+	c.its = c.its[:0]
+	c.h = c.h[:0]
+	c.batches = c.batches[:0]
+
+	if len(c.its) <= cap(c.batchesBuf) {
+		c.batchesBuf = c.batchesBuf[:size]
+		for i := 0; i < size; i++ {
+			c.batchesBuf[i] = promchunk.Batch{}
```

Comment:
I wonder if `promchunk.Batch{}` is allocated in heap or not... Is it cheaper to reset a batch using code below?

```
	batch.Length = 0
	batch.Index = 0
```

Reply:
I dont think the new struct scapes to the heap (as it is local and we are just copying the values to the struct already created on the slice: 

See the results with the proposed change:

```
goos: darwin
goarch: arm64
pkg: github.com/cortexproject/cortex/pkg/querier/batch
cpu: Apple M1 Pro
BenchmarkMergeIterator
BenchmarkMergeIterator/reuse-true
BenchmarkMergeIterator/reuse-true-10         	   13790	     88019 ns/op	    7672 B/op	      46 allocs/op
BenchmarkMergeIterator/reuse-false
BenchmarkMergeIterator/reuse-false-10        	   13404	     88980 ns/op	   16680 B/op	      51 allocs/op
```

I did this way to be sure we are reseting all fields (seems a bit safer?)